### PR TITLE
Add BatchingHttpClient

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Refs https://github.com/elifesciences/api-sdk-php/pull/78.

### CODE REVIEW

File: src/HttpClient/BatchingHttpClient.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public function __construct(HttpClient $httpClient, int $batchSize = 10)
+    {
+        $this->httpClient = $httpClient;
+        $this->batchSize = $batchSize;
+    }
+
+    public function send(RequestInterface $request) : PromiseInterface
+    {
+        $this->ifBatchIsFull(function () {
+            $this->cleanBatchList();
```
Comment (stephenwf) [inline src/HttpClient/BatchingHttpClient.php]: Does `$this` would in the class scope?

File: src/HttpClient/BatchingHttpClient.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        });
+        $this->ifBatchIsFull(function () {
+            $this->waitOnBatch();
+        });
+
+        return $this->batch[] = $this->httpClient->send($request);
+    }
+
+    private function cleanBatchList()
```
Comment (giorgiosironi) [inline src/HttpClient/BatchingHttpClient.php]: sounds more like `cleanBatchListFromCompletedRequests`

**Comment by giorgiosironi** [comment]
Needs some extensive testing, we can introduce some randomized input?

**Comment by thewilkybarkid** [comment]
Yep, do you want to sort it?

File: src/HttpClient/BatchingHttpClient.php (hunk 2 above)
Comment (thewilkybarkid) [inline src/HttpClient/BatchingHttpClient.php]: How about `filterResolvedRequests`? (Some may have failed.)

File: src/HttpClient/BatchingHttpClient.php (hunk 1 above)
Comment (thewilkybarkid) [inline src/HttpClient/BatchingHttpClient.php]: Yep, as of PHP 5.4.

**Comment by giorgiosironi** [comment]
Ok, will pull request on the pull request.

**Comment by thewilkybarkid** [comment]
Just push to it

**Comment by giorgiosironi** [comment]
The randomized test hits the client with lots of requests alternated with waits on the returned promises, checking responses come out as they should. Check the commit message for details.

**Comment by giorgiosironi** [comment]
`solves` is better, didn't have a term for both completed and failed requests

**Comment by thewilkybarkid** [comment]
'resolved' covers both

File: composer.json [hunk 3]
```diff
@@ -28,7 +28,8 @@
         "phpspec/phpspec": "^2.4",
         "phpspec/prophecy": "^1.5",
         "phpunit/phpunit": "^5.2",
-        "psr/log": "^1.0"
+        "psr/log": "^1.0",
+        "giorgiosironi/eris": "^0.8.0"
```
Comment (thewilkybarkid) [inline composer.json]: Alphabetical please. (Just `^0.8` is fine too.)

Comment (giorgiosironi) [inline composer.json]: ok, this was added by `composer require --dev`

**Comment by giorgiosironi** [comment]
I wrongly pushed to `origin` 10/15 times, but now it's green.

**Comment by thewilkybarkid** [comment]
Good to merge?

**Comment by giorgiosironi** [comment]
Good for me, next up is integrating it in api-sdk and other projects.

**Pull request merged**